pkg/apis/k0s.k0sproject.io/v1beta1/clusterconfig_types.go

```diff
@@ -277,45 +282,50 @@ func DefaultClusterSpec(defaultStorage ...*StorageSpec) *ClusterSpec {
 	}
 }
 
-func (c *ControllerManagerSpec) Validate() []error {
-	return nil
+// Validateable interface to ensure that all config components implement Validate function
+// +k8s:deepcopy-gen=false
+type Validateable interface {
+	Validate() []error
 }
 
-var _ Validateable = (*SchedulerSpec)(nil)
+func (s *ClusterSpec) Validate() (errs []error) {
+	if s == nil {
+		return
+	}
+
+	for name, field := range map[string]Validateable{
+		"api":               s.API,
+		"controllerManager": s.ControllerManager,
+		"scheduler":         s.Scheduler,
+		"storage":           s.Storage,
+		"network":           s.Network,
+		"workerProfiles":    s.WorkerProfiles,
+		"telemetry":         s.Telemetry,
+		"install":           s.Install,
+		"extensions":        s.Extensions,
+		"konnectivity":      s.Konnectivity,
+	} {
```

Comment:
Would it be overkill to use reflection here to loop over all the fields instead of manually maintaining this list?

Reply:
I think the code wouldn't look nice. There's a lot of possible errs that need to be checked, and there's potential for panics. Eventually it might be worthwhile to have a look at kubebuilder. I know for example that it can handle the defaulting stuff that's currently done done manually in k0s. Maybe there's something in there to generate validations, too.

Reply:
A cheap solution to make sure all fields are in the list is to add a test that checks for this using reflection.

Reply:
...so we would have both reflection and manually maintained list at the same time :trollface:

Reply:
Precisely my point 😄
But jokes aside, the code would be easy to read and follow, and a test will keep us from forgetting new fields.

Reply:
If we want to use reflection, this would be enough:
```
val := reflect.ValueOf(s).Elem()
for pos := 0; pos <= val.NumField() - 1; pos++ {
	tag, ok := val.Type().Field(pos).Tag.Lookup("json")
	if ok {
		for _, err := range val.Field(pos).Interface().(Validateable).Validate() {
			errs = append(errs, fmt.Errorf("%s: %w", tag, err))
		}
	}
}
```
But I'm not still sure that we need it